### DISCUSSION TITLE
CLR extension methods

### DISCUSSION THREAD

**Discussion by KoalaBear84** [sebastienros/jint#1067] [Q&A] [answered]
Saw a question on [Stack Overflow](https://stackoverflow.com/questions/63939864/jint-using-clr-object-s-properties-in-javascript-functions/70875893#70875893) while researching and couldn't get it to work with the extension methods provided by C# itself.

```csharp
// Works
public class Foo
{
    public string Name { get; } = "Bar";
}

public static class CustomStringExtensions
{
    public static bool StartsWith(this string value, string value2) => value.StartsWith(value2);
}

Engine engine = new Engine(options =>
{
    options.AddExtensionMethods(typeof(CustomStringExtensions));
});

bool result = engine
    .SetValue("foo", new Foo())
    .Evaluate("foo.Name.StartsWith('B')")
    .AsBoolean();
```

```csharp
// Doesn't work
public class Foo
{
    public string Name { get; } = "Bar";
}

Engine engine = new Engine(options =>
{
	// Also tried these
	//options.AllowClr();
	//options.AllowOperatorOverloading();
	//options.AddExtensionMethods(typeof(String));
	options.AddExtensionMethods(typeof(string));
});

bool result = engine
	.SetValue("foo", new Foo())
	.Evaluate("foo.Name.StartsWith('B')")
	.AsBoolean();

bool result = engine
    .SetValue("foo", new Foo())
    .Evaluate("foo.Name.StartsWith('B')")
    .AsBoolean();
```

Is it because some overloading? I'm curious about the reason.

I see there isn't really anything documented about how the extension method support is to be used. Could an example be added, and maybe also that native extensions methods do not work, or if they do, how they can be added. Thanks!

When will version 3 become available, as I see it's already beta since Oct 2016? :)

**Comment by lahma** [comment, marked as answer]
`options.AddExtensionMethods(typeof(string));` does not make sense as `String` doesn't (cannot) contain any extension methods, they reside in `static` classes.

Your code should work if you write it as `foo.Name.startsWith('B')`, Jint matches using lower-case first letter by default to follow JS naming. You can override this via `options.TypeResolver.MemberNameComparer` configuration.

Best place to learn about extension methods is to read [the related tests](https://github.com/sebastienros/jint/tree/main/Jint.Tests/Runtime/ExtensionMethods) I guess.

Version 3 __is available__ as beta builds from NuGet, don't know when it hits final stage.

**Comment by KoalaBear84** [comment]
Thanks! Ahh, sorry, indeed, those aren't extension methods, although they feel like it.
